### PR TITLE
[core] Use primary bytes when calculating when to preemptively spill

### LINKED ISSUE #36044: [core] Primary copies are spilled when secondary copies could have been evicted
When running workloads that read secondary copies onto a node, Ray seems to sometimes spill primary copies on the node before it evicts secondary copies on the node. This leads to unnecessary spilling, since evicting secondary copies doesn't spill.

The following can reproduce the issue on a 6-node cluster (with total ~200GB object store memory, plenty to fit all the objects without spilling):

```
import ray
import numpy as np

# Spread data onto all nodes.
@ray.remote
def gen():
    return np.ones(1024 * 1024 * 1024, dtype=np.uint8)

refs = [gen.options(scheduling_strategy="SPREAD").remote() for _ in range(50)]  # 50GiB

# Iterate over the data on the worker nodes from the head node.
for i in range(10):
    for j, r in enumerate(refs):
        print("Iteration", i, j)
        ray.get(r)

```

Screenshot from 2023-06-02 14-56-18

You can see that a small fraction of the memory is spilled, which is not necessary. The spilling doesn't happen if you prevent the tasks from running on the head node, suggesting that the spilling is only due to primary copies created on that node.

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Ray uses a threshold to decide when to preemptively spill, to avoid pauses when object spilling is triggered. However, this threshold was previously applying to both primary and secondary objects, which led to unnecessary spilling. This PR moves the logic to a periodic check in the NodeManager and uses the primary bytes instead of all allocated bytes.

## Related issue number

Closes #36044.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
This may cause some performance degradation in spill-based benchmarks (like dataset shuffle), but we can adjust the threshold if that happens.

I'm also going to remove the dead code in plasma/store.cc but this can be reviewed first.

File: src/ray/raylet/node_manager.cc
Comment (jjyao) [inline src/ray/raylet/node_manager.cc:529]:
```suggestion
            RAY_LOG(INFO) << "Triggering object spilling because current primary objects usage "
```

File: src/ray/raylet/node_manager.cc
Comment (jjyao) [inline src/ray/raylet/node_manager.cc:522]: Why is it better to spill here than the previous `HandleCreateObjectRequest`?

Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:522]: I needed access to the number of bytes held by primary objects. LocalObjectManager has the ground truth.

File: src/ray/raylet/node_manager.cc
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:536]: The previous code seems to have a return value

```
return GetLocalObjectManager().IsSpillingInProgress();
```

And now the store.cc cannot catch this state. Would there be any behavior change because of this? Looking at code, it seems like it shouldn't matter (seems like spill callback that's used by `spilling_required` doesn't use the return value), but I'd like to double check.

Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:536]: Do you know how frequently we triggered threashold-based spilling before this? Now it seems like we are doing once a second. Was it the same, or more frequent? Is it possible to preserve the same interval & frequency to minimze the impact of changes>?

**Review by rkooo567 — COMMENTED** [review]
Code LGTM. Two questions regarding the behavior changes.

File: python/ray/tests/test_object_spilling_3.py
Comment (rkooo567) [inline python/ray/tests/test_object_spilling_3.py:367]: nit: add a docstring to the test

"""
Verify spilling is not triggered when secondary copies are evictable
"""

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:536]: I don't think there is any behavior change.

Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:536]: Before it was once per object. Good idea; I'll add the same check to the object added handler.

File: python/ray/tests/test_object_spilling_3.py
Comment (stephanie-wang) [reply python/ray/tests/test_object_spilling_3.py:367]: I think this is fine because the test name already says this.

**Review by rkooo567 — APPROVED** [review]
LGTM given https://github.com/ray-project/ray/pull/38805/files#r1304426892 is addresed

**Comment by zhe-thoughts** [comment]
@stephanie-wang could you help me understand why this is a release blocker (is it OK to include it in 2.8 instead?)

**Comment by stephanie-wang** [comment]
> @stephanie-wang could you help me understand why this is a release blocker (is it OK to include it in 2.8 instead?)

I don't think it should be a release blocker (there's no regression) but it is a bug in spilling behavior and was affecting some Data user workloads. I'll defer to @c21 about the latter.

**Comment by c21** [comment]
Yeah, agree this is not a release blocker, more context is in https://github.com/ray-project/ray/issues/36044#issuecomment-1692044309 .

**Comment by zhe-thoughts** [comment]
OK, removing myself as an assigning, and let's merge this after Friday 5pm

**Pull request merged**
